src/libertem/executor/pipelined.py

```diff
@@ -0,0 +1,857 @@
+
+        # timeout for starting a single worker
+        self._startup_timeout = startup_timeout
+
+        # keep this at the bottom:
+        self._pool = self._start_pool()
+
+    def _start_pool(self) -> WorkerPool:
+        with tracer.start_as_current_span("PipelinedExecutor.start_pool") as span:
+            pool = WorkerPool(
+                worker_fn=functools.partial(
+                    pipelined_worker,
+                    pin=self._pin_workers,
+                    early_setup=self._early_setup,
+                ),
+                spec=self._spec,
+            )
+            # FIXME: check process liveness here, too, to catch early startup
+            # failures without having to run into the timeout
+            for i in range(pool.size):
+                with pool.response_queue.get(timeout=self._startup_timeout) as (msg, _):
+                    if msg["type"] == "ERROR":
+                        raise RuntimeError(
+                            f"error on startup: {msg['error']}"
+                        )
+                    if msg["type"] != "STARTUP_DONE":
+                        raise RuntimeError(
+                            f"unknown message type {msg['type']}, expected STARTUP_DONE"
+                        )
+                    span.add_event("worker startup done", {"worker_id": msg["worker_id"]})
+            # set here, so we don't try to close the pool if it doesn't exist
+            self._closed = False
+            return pool
+
+    @classmethod
+    def _default_spec(cls):
+        from libertem.utils.devices import detect
+        detected = detect()
+        return _make_spec(**detected)
+
+    @classmethod
+    def make_local(cls, **kwargs):
+        """
+        Create a :code:`PipelinedExecutor` with the default spec.
+        """
+        spec = cls._default_spec()
+        return cls(spec=spec, **kwargs)
+
+    @classmethod
+    def make_spec(cls, *args, **kwargs):
+        return _make_spec(*args, **kwargs)
+    make_spec.__doc__ = _make_spec.__doc__
+
+    def _validate_worker_state(self):
+        if not self._pool.all_alive():
+            raise RuntimeError("some workers are stopped, cannot continue")
+
+    def _run_tasks_inner(
+        self,
+        tasks: Iterable[TaskProtocol],
+        params_handle: Any,
+        cancel_id: Any,
+        task_comm_handler: "TaskCommHandler",
+    ) -> ResultWithID:
+        in_flight = 0
+        id_to_task = {}
+
+        try:
+            self._validate_worker_state()
+            task_comm_handler.start()
+            span = trace.get_current_span()
+            span_context = span.get_span_context()
+
+            for qs, _ in self._pool.workers:
+                qs.request.put({
+                    "type": "WARMUP",
+                    "span_context": span_context,
+                })
+
+            for task_idx, task in enumerate(tasks):
+                in_flight += 1
+                id_to_task[task_idx] = task
```

Comment:
Good points all. I would imagine some kind of system where each 'job' (udf task or other) has a uuid associated, and results are pulled from the queue or an intermediate cache by requesting the particular UUID. In truth this is essentially re-implementing `concurrent.futures` or `Dask.futures` and we might want to look if we can re-use some of that API one day?

Reply:
>  I would imagine some kind of system where each 'job' (udf task or other) has a uuid associated, and results are pulled from the queue or an intermediate cache by requesting the particular UUID. 

Yeah exactly.

> In truth this is essentially re-implementing `concurrent.futures` or `Dask.futures` and we might want to look if we can re-use some of that API one day?

Agreed, using API from `concurrent.futures` would be really nice. I think the tricky part here is the synchronization between starting a task and sending the data to the correct queue - right now, this is handled by sharing a queue for synchronization (starting a task etc.) and sending data.

It _could_ be possible to re-model this to a pull-model, meaning when a task is ready for execution, it specifically calls back to the main process and requests the data for the given task, which then blocks until that data is available. I think with that model, it would be easier to re-use any kind of process pool or `concurrent.futures.Executor` kind of abstraction - on the other hand, there would have to be additional buffering on the main process, keeping the data around until it can be sent to the correct queue.

_Then_, if this buffering happens in shared memory already, it _could_ work (at least) as well as it does now - as long as this is a single-node executor - in the distributed case, it would really be nice to speculatively send data to its destination as soon as we get it on the main node. Kind of like the difference between a cut-through switch and one that uses store-and-forward.

With `dask.futures` the main issue is we need an "out-of-band" communication channel, that matches the IPC that is used for client<>scheduler<>worker communication.